documents/design/FirmwareRefactor.md

```diff
@@ -0,0 +1,190 @@
+
+This will also be the starting point for the other threads. Ideally, the other threads will be started before establishing the TCP socket. The other threads would run idly in the background until necessary since the TCP socket could be reset if connection is lost.
+
+#### Motor Control
+
+The motor control thread will contain the code for the PID control loop. It will use a `MotorController` class (described below) to send commands to the motor controller. It will also keep track of the encoder ticks. This thread is fairly simple, as it will likely reuse a lot of existing code.
+
+#### Diagnostics
+
+The diagnostics thread will contain the code for communicating with the CAN devices and diagnostics board. It will be in charge of monitoring those devices and taking appropriate actions when necessary. The main function of the thread is constructing messages to send back to the computer for bagging and debugging. The first step in implementing this will be defining what the devices are and how they work. In the future, wrapper classes for the devices and diagnostics board may be useful. This code will change the most frequently in the coming year.
+
+---
+
+### Code Structure
+
+Currently, the firmware follows no specific structure. All code is kept in one `main.cpp` file and follows no overall design pattern. The overall code structure will be fixed with the changes described in the threading section, but this section will propose some changes that will help justify those changes and will make implementing the threads easier. Most of these changes will remove code from the `main.cpp` file into additional header/cpp files.
+
+#### `MotorController`
+
+A `MotorController` wrapper class will be created to handle communicating with the motor controller and will abstract its functionality. This change will accomplish 2 things: declutter the code for sending commands to the motor and achieve the overall software goal of writing self-documenting code. Here is how the motor controller is currently called in the code:
+
+``` Serial g_serial(p13, NC, 9600); ```
+
+This does nothing to explain what it is or how it works. A similar issue can be found in the PID control loop. The latter half of the `pid()` method is dedicated to converting the speed into motor values. This code should be taken out of that method and moved to the MotorController class. Here are the base functions that the `MotorController` class should provide:
+
+- Set the speeds of left and right motors independently
+- Get the speeds sent to the motors
+- Stop the motors
+- Advanced control of the motors (slowly stop the motors, turn around, etc.)
```

Comment:
Not too sure about this one, as I don't see anything that will make use of this

Reply:
I put that there more as an idea for potential uses of the class beyond just basic functionality. I don't expect us to do anything like that unless we were going to do something very fancy with controls.